ce/api/streamflow/downstream.py

```diff
@@ -0,0 +1,163 @@
+    :param flow_direction: (VicDataGrid) Flow direction grid
+    :return: (dict) representation for JSON response object; see downstream() for details
+    """
+
+    # Compute lonlats of watershed whose starting point is `station`
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_downstream_watershed(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    watershed_lonlats = [flow_direction.xy_to_lonlat(xy) for xy in watershed_xys]
+
+    # Compute outline of watershed as a GeoJSON feature
+    outline = path_line(watershed_lonlats)
+
+    return {
+        "boundary": geojson_feature(
+            outline,
+            properties={
+                "starting point": geojson_feature(
+                    Point(
+                        flow_direction.xy_to_lonlat(
+                            flow_direction.lonlat_to_xy(station_lonlat)
+                        )
+                    ),
+                )
+            },
+        ),
+    }
+
+
+def build_downstream_watershed(target, routing, direction_map, debug=False):
+    """
+    Return set of all cells (including target) that drain from `target`.
+
+    :param target: An xy index representing the cell of interest.
+    :param routing: A numpy array representing water flow using VIC direction
+        codes (0 - 9). These codes may be integers or floats.
+    :param direction_map: Maps VIC direction codes into offsets in cell indices.
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Tuple of cells (cell indices as tuples) that drain from `target` in 
+        a downstream flow order.
+
+    Notes:
+
+    - In this function, a cell is represented by an (x, y) index pair.
+
+    - Routing graphs can and in practice do contain cycles. Variable 
+    `downstream tuple` is used to check whether a cell has already been 
+    visited during the traversal of the routing graph, i.e., whether we 
+    are cycling, and if so not to repeat that subgraph.
+    """
+
+    def downstream(downstream_tuple):
+        if direction_map == None:
+            return downstream_tuple
+
+        cell_routing = routing[downstream_tuple[-1]]
+        downstream_neighbour = vec_add(downstream_tuple[-1], direction_map[int(cell_routing)])
+
+        if downstream_neighbour in downstream_tuple or not is_valid_index(downstream_neighbour, routing.shape):
+            return downstream_tuple
+
+        downstream_tuple += (downstream_neighbour,)
+        return downstream(downstream_tuple)
```

Comment:
That's very nice. I like how you've rolled `visited` into `downstream_tuple`. I also like how you are using the last element of `downstream_tuple` as the starting point of each recursion. 

My only suggestion here would be to find a more elegant name for `downstream_tuple` -- perhaps `stream`? In general, it's best to avoid putting the data type (in this case, tuple) in the name; instead just choose a name that conveys its purpose or meaning. Amongst other things, you might change the datatype later, even if its meaning is the same.

Reply:
Aside: This is what's known as a tail recursion, and it's the recursive equivalent of a loop in procedural programming. 

You could transform this algorithm into a loop, but there's no reason to; recursion is fine too. Some compilers actually recognize tail recursion and compile to loop code, which is generally more efficient in space and time (by a constant factor, not a fundamental algorithmic difference like linear time vs quadratic time). Not sure if Python does that.